### PR TITLE
internal/extsvc/github: Support field Visibility for GetRepository

### PR DESCRIPTION
**Pull request by indradhanush**
Required by #25904.

### CODE REVIEW

File: internal/extsvc/github/common.go [hunk 1]
```diff
@@ -1660,6 +1660,9 @@ type Repository struct {
 	// Metadata retained for ranking
 	StargazerCount int `json:",omitempty"`
 	ForkCount      int `json:",omitempty"`
+
+	// public or private or internal.
```
Comment (eseliger) [inline internal/extsvc/github/common.go:1664]: Can you add a comment this is only set when using GetRepository?

Comment (indradhanush) [reply internal/extsvc/github/common.go]: I digged a bit further and it seems it is not just restricted to `GetRepository` according to [this](https://developer.github.com/changes/2019-12-03-internal-visibility-changes/#repository-visibility-fields). 

I've added some explanation and this link to the doc.

File: internal/extsvc/github/common_test.go [hunk 2]
```diff
@@ -84,7 +84,8 @@ func TestClient_GetRepository(t *testing.T) {
 	"fork": true,
 	"stargazers_count": 30,
 	"watchers_count": 20,
-	"forks_count": 5
+	"forks_count": 5,
+    "visibility": "internal"
```
Comment (ryanslade) [inline internal/extsvc/github/common_test.go:88]:
```suggestion
   "visibility": "internal"
```

File: internal/extsvc/github/common.go (hunk 1 above)
Comment (eseliger) [reply internal/extsvc/github/common.go]: but we don't do it for GraphQL

Comment (indradhanush) [reply internal/extsvc/github/common.go]: @eseliger I updated the comment and added another one where we add the request header. Let me know if I can improve it in any way and make it more explicit.

File: internal/extsvc/github/common.go [hunk 3]
```diff
@@ -1660,6 +1660,11 @@ type Repository struct {
 	// Metadata retained for ranking
 	StargazerCount int `json:",omitempty"`
 	ForkCount      int `json:",omitempty"`
+
+	// This is available for GitHub Enterprise Cloud and GitHub Enterprise Server 2.20+ and is used
+	// to identify if a repository is public or private or internal.
+	// https://developer.github.com/changes/2019-12-03-internal-visibility-changes/#repository-visibility-fields
+	Visibility Visibility `json:",omitempty"`
```
[the remark this answers is not in the document]
Comment (eseliger) [reply internal/extsvc/github/common.go:1667]: I think you want to capture visibility in GraphQL as well:
image

https://docs.github.com/en/graphql/reference/objects#repository

[the remark this answers is not in the document]
Comment (ryanslade) [reply internal/extsvc/github/common.go:1667]: Yep, you need to make sure we're including the field there too. I think you just need to add it here:
https://github.com/sourcegraph/sourcegraph/blob/a8b56db8b523045179921b2b988813009f3a8fe7/internal/extsvc/github/v4.go#L539

Also, it looks like it comes back uppercase from the GraphQL API so you may need to convert it

[the remark this answers is not in the document]
Comment (eseliger) [reply internal/extsvc/github/common.go:1667]: Also, when was this introduced in GitHub? The request will fail on instances that don't know about that field so it won't be backwards compatible. If not too long ago, I think we need to feature flag including it. (There's other cases for that too, we got the heavy lifting for it already so not as bad as it sounds I hope)

[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/github/common.go:1667]: It was introduced in December 2019 as of this blog: https://developer.github.com/changes/2019-12-03-internal-visibility-changes/#repository-visibility-fields

But this says: 

> Customers with an enterprise account using GitHub Enterprise Cloud and GitHub Enterprise Server 2.20+ have access to a third visibility option beyond public and private, called internal. We've made some recent API changes to support this option.

So I'm not a entirely sure when the `visibility` field itself was added to the API.

File: internal/extsvc/github/common_test.go [hunk 4]
```diff
@@ -84,7 +84,8 @@ func TestClient_GetRepository(t *testing.T) {
 	"fork": true,
 	"stargazers_count": 30,
 	"watchers_count": 20,
-	"forks_count": 5
+	"forks_count": 5,
+  "visibility": "internal"
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/github/common_test.go:88]: Argh. :grimacing: 

I ran a prettier on the entire JSON.

File: internal/extsvc/github/common.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanslade) [reply internal/extsvc/github/common.go:1667]: It looks like we have code to add fields depending on the version here: 
https://github.com/sourcegraph/sourcegraph/blob/main/internal/extsvc/github/v4.go#L543-L547

**Review by ryanslade — APPROVED** [review]

**Review by eseliger — COMMENTED** [review]
Can you please add a test to validate that it is returned correctly from the graphql API? I am almost sure the GraphQL value will be UPPERCASE, while the rest value seems to be lowercase, so we probably need to align those.

**Comment by indradhanush** [comment]
@eseliger So our internal GHE instance is running 2.2.0 at the moment. Without having a reliable way to test this, we decided to guard this work for both the REST and the GraphQL API behind a feature flag (disabled by default). 

In the meantime, we will be working to figure out a way to upgrade to GHE 3.3.0 and then we can come back to enable the feature flag and test locally - making any changes that might require and bring another PR in. :smile:

File: internal/extsvc/github/common.go [hunk 5]
```diff
@@ -1660,6 +1662,11 @@ type Repository struct {
 	// Metadata retained for ranking
 	StargazerCount int `json:",omitempty"`
 	ForkCount      int `json:",omitempty"`
+
+	// This is available for GitHub Enterprise Cloud and GitHub Enterprise Server 2.20+ and is used
```
Comment (eseliger) [inline internal/extsvc/github/common.go:1666]:
```suggestion
	// This is available for GitHub Enterprise Cloud and GitHub Enterprise Server 3.3.0+ and is used
```

**Review by eseliger — APPROVED** [review]
🚀

**Review by ryanslade — APPROVED** [review]

**Pull request merged**